services/logs/filter.go

```diff
@@ -0,0 +1,216 @@
+		start:    &start,
+		end:      &end,
+		criteria: criteria,
+		receipts: receipts,
+	}
+}
+
+func (r *RangeFilter) Match() ([]*gethTypes.Log, error) {
+	if r.start.Cmp(r.end) != -1 {
+		return nil, fmt.Errorf("invalid start and end block height, start must be smaller than end value")
+	}
+
+	blooms, heights, err := r.receipts.BloomsForBlockRange(r.start, r.end)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(blooms) != len(heights) {
+		return nil, fmt.Errorf("bloom values don't match height values") // this should never happen
+	}
+
+	logs := make([]*gethTypes.Log, 0)
+	for i, bloom := range blooms {
+		if !bloomMatch(bloom, r.criteria) {
+			continue
+		}
+
+		// todo do this concurrently
+		receipt, err := r.receipts.GetByBlockHeight(heights[i])
+		if err != nil {
+			return nil, err
+		}
+
+		for _, log := range receipt.Logs {
+			if exactMatch(log, r.criteria) {
+				logs = append(logs, log)
+			}
+		}
+	}
+
+	return logs, nil
+}
+
+// IDFilter matches all logs against the criteria found in a single block identified
+// by the provided block ID.
+type IDFilter struct {
+	id       common.Hash
+	criteria FilterCriteria
+	blocks   storage.BlockIndexer
+	receipts storage.ReceiptIndexer
+}
+
+func NewIDFilter(
+	id common.Hash,
+	criteria FilterCriteria,
+	blocks storage.BlockIndexer,
+	receipts storage.ReceiptIndexer,
+) *IDFilter {
+	return &IDFilter{
+		id:       id,
+		criteria: criteria,
+		blocks:   blocks,
+		receipts: receipts,
+	}
+}
+
+func (i *IDFilter) Match() ([]*gethTypes.Log, error) {
+	blk, err := i.blocks.GetByID(i.id)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := i.receipts.GetByBlockHeight(big.NewInt(int64(blk.Height)))
+	if err != nil {
+		return nil, err
+	}
+
+	logs := make([]*gethTypes.Log, 0)
+	for _, log := range receipt.Logs {
+		if exactMatch(log, i.criteria) {
+			logs = append(logs, log)
+		}
+	}
+
+	return logs, nil
+}
+
+// StreamFilter matches all the logs against the criteria from the receipt channel.
+type StreamFilter struct {
+	criteria      FilterCriteria
+	receiptStream chan *gethTypes.Receipt
+}
+
+func NewStreamFilter(criteria FilterCriteria, receipts chan *gethTypes.Receipt) *StreamFilter {
+	return &StreamFilter{
+		criteria:      criteria,
+		receiptStream: receipts,
+	}
+}
+
+func (s *StreamFilter) Match() (chan *gethTypes.Log, error) {
+	logs := make(chan *gethTypes.Log)
+
+	go func() {
+		defer close(logs)
+
+		for {
+			select {
+			case receipt, ok := <-s.receiptStream:
+				if !ok {
+					return // exit the goroutine if receiptStream is closed
+				}
+
+				if !bloomMatch(receipt.Bloom, s.criteria) {
+					continue
+				}
+
+				for _, log := range receipt.Logs {
+					if exactMatch(log, s.criteria) {
+						logs <- log
+					}
+				}
+			}
+		}
+	}()
+
+	return logs, nil
+}
+
+// exactMatch checks the topic and address values of the log match the filer exactly.
+func exactMatch(log *gethTypes.Log, criteria FilterCriteria) bool {
+	if len(criteria.Topics) > len(log.Topics) {
```

Comment:
should this be?
```suggestion
	if len(criteria.Topics) != len(log.Topics) {
```

Reply:
It shouldn't because the criteria can have less topics than the log, in case where topics are wildcards.

Reply:
added a comment